addons/sale_coupon/tests/test_program_with_code_operations.py

```diff
@@ -379,3 +379,57 @@ def test_apply_program_no_reward_link(self):
         }).process_coupon()
         self.assertFalse(self.empty_order.applied_coupon_ids, 'No coupon should be linked to the order')
         self.assertEqual(coupon.state, 'new', 'Coupon should be in a new state')
+
+    def test_delete_all_discount_lines(self):
+        """
+            The goal is to ensure that all discount lines are deleted
+            when we need to update existing reward lines.
+        """
+        self.env['coupon.program'].create({
+            'name': '50% Discount on order',
+            'promo_code_usage': 'code_needed',
+            'promo_code': 'test',
+            'reward_type': 'discount',
+            'discount_type': 'percentage',
+            'discount_percentage': 50,
+            'active': True,
+            'discount_apply_on': 'on_order',
+        })
+        product_with_tax, product_without_tax = self.env['product.product'].create([
+            {
+                'name': 'Product with tax',
+                'list_price': 100,
+                'sale_ok': True,
+                'taxes_id': [self.tax_10pc_excl.id],
+            },
+            {
+                'name': 'Product without tax',
+                'list_price': 50,
+                'sale_ok': True,
+                'taxes_id': [],
+            }
+        ])
+        order = self.empty_order.copy()
+        order.write({'order_line': [
+            (0, False, {
+                'product_id': product_with_tax.id,
+                'name': '1 Product with tax',
+                'product_uom': self.uom_unit.id,
+                'product_uom_qty': 1.0,
+            }),
+            (0, False, {
+                'product_id': product_without_tax.id,
+                'name': '1 Product without tax',
+                'product_uom': self.uom_unit.id,
+                'product_uom_qty': 1.0,
+            })
+        ]})
+        self.env['sale.coupon.apply.code'].with_context(active_id=order.id).create({
+            'coupon_code': 'test'
+        }).process_coupon()
+        self.assertEqual(len(order.order_line), 4, '2 products and 2 discount lines')
+        line_to_remove = order.order_line.filtered(lambda l: l.name == '1 Product without tax')
+        order.write({'order_line': [(3, line_to_remove.id, 0)]})
+        order.recompute_coupon_lines()
+        reward_lines = order.order_line.filtered(lambda l: l.is_reward_line)
+        self.assertFalse(reward_lines, 'All reward lines must be removed')
```

Comment:
```suggestion
        self.assertEqual(order.amount_total, 160)
        self.env['sale.coupon.apply.code'].with_context(active_id=order.id).create({
            'coupon_code': 'test'
        }).process_coupon()
        self.assertEqual(order.amount_total, 80)
        self.assertEqual(len(order.order_line), 4, '2 products and 2 discount lines')
        line_to_remove = order.order_line.filtered(lambda l: l.name == '1 Product without tax')
        order.write({'order_line': [(3, line_to_remove.id, 0)]})
        order.recompute_coupon_lines()
        self.assertEqual(order.amount_total, 55)  # Works before your patch, not after
        reward_lines = order.order_line.filtered(lambda l: l.is_reward_line)
        self.assertFalse(reward_lines, 'All reward lines must be removed')
```

I tried to follow your test and see the difference before/after your patch, and I notice that you're basically modifying the existing behavior of coupons application.
I would expect the discount to still be applied after I remove one of the discounted lines (unless there is a specific condition not working anymore), but it's not after your changes.